File: src/solar/solar-common/util/ability/AbilityButtonUtil.ts
import FrameCallbackUtil from "@/util/frame/FrameCallbackUtil";
import AbilityUtil from "./AbilityUtil";
import KeyCode from "@/constant/KeyCode";
import LangUtil from "@/LangUtil";
import FramePoint from "@/FramePoint";


export default class AbilityButtonUtil {

    static qwerKeyCodes: number[] = [KeyCode.VK_Q, KeyCode.VK_W, KeyCode.VK_E, KeyCode.VK_R];
    static qwerKeyNames: string[] = ["Q", "W", "E", "R"];
    /**
     *
     *  只能设置的按钮位置 为ture时  不会被没有设置位置的顶用
     * （以防止 设置技能的位置时  后面的未设置位置的 跑到前面的空位置去了）
     * 如果只有 WE 技能可设置 则buttonFixeds = [false, true, true, false];
     *
     */
    static onlySetButtons = [false, false, false, false];
    private static _sl_bottomButtonArray: number[] = null;

    /**
     * 整数技能的id  与技能位置的 映射关系
     * 技能排序时 可以从这里获得技能的位置
     *
     * 注意这个数据是异步的 每个玩家都不一样
     *
     */
    static abilityIdButtonXMap: {
        [aId: number]: number
    } = {}

    /**
     * 同步的技能数据 可做同步计算
     */
    static syncAbilityIdButtonXMap: {
        [aId: number]: { [aId: number]: number }
    } = {}

    /**
     * 设置技能底部按钮x位置
     * 只能QWER使用 本地玩家使用
     *
     * 注意单位 不能拥有相同位置的技能 否则会重叠按钮
     * @param unit
     * @param abilityIdStr
     * @param x
     */
    static setAbilityBottomButtonXAndHotKey(unit: unit, abilityIdStr: string, x: number) {
        let player = GetOwningPlayer(unit)
        let keyCode = AbilityButtonUtil.qwerKeyCodes[x]
        AbilityUtil.setUnitAbilityHotkey(unit, abilityIdStr, keyCode, false)
        AbilityUtil.setUnitAbilityName(unit, abilityIdStr, _g_objs.ability[abilityIdStr].Name + "(" + AbilityButtonUtil.qwerKeyNames[x] + ")", true)
        return AbilityButtonUtil.setAbilityBottomButtonX(abilityIdStr, x, player)
    }

    /**
     * 设置技能底部按钮x位置
     * 只能QWER使用 本地玩家使用
     *
     * 注意单位 不能拥有相同位置的技能 否则会重叠按钮
     * @param abilityIdStr
     * @param x
     * @param player
     */
    static setAbilityBottomButtonX(abilityIdStr: number | string, x: number, player: player) {
        abilityIdStr = LangUtil.getIntId(abilityIdStr);
        let sABXMPD = AbilityButtonUtil.syncAbilityIdButtonXMap[GetPlayerId(player)];
        if (!sABXMPD) {
            sABXMPD = {};
            AbilityButtonUtil.syncAbilityIdButtonXMap[GetPlayerId(player)] = sABXMPD;
        }
        sABXMPD[abilityIdStr] = x
        if (player != GetLocalPlayer()) {
            return
        }
        AbilityButtonUtil._sl_init_bottomButton_system();
        AbilityButtonUtil.abilityIdButtonXMap[abilityIdStr] = x
    }

    private static _sl_isInitialized = false

    private static _sl_init_bottomButton_system() {
        if (AbilityButtonUtil._sl_isInitialized) {
            return;
        }
        AbilityButtonUtil._sl_bottomButtonArray = [
            DzFrameGetCommandBarButton(2, 0),
            DzFrameGetCommandBarButton(2, 1),
            DzFrameGetCommandBarButton(2, 2),
            DzFrameGetCommandBarButton(2, 3),
        ]
        for (let index = 0; index < 4; index++) {
            const uiid = AbilityButtonUtil._sl_bottomButtonArray[index]
            DzFrameClearAllPoints(uiid)
        }
        AbilityButtonUtil._sl_resetBottomButton();
        FrameCallbackUtil.addFrameSetUpdateCallback(AbilityButtonUtil._sl_refreshButton);
        AbilityButtonUtil._sl_isInitialized = true
    }

    private static lastFrameSelection = null;
    private static lastNoAbility = false

    private static _sl_refreshButton(this: void) {
        //没有自定义的qwer 则可能选择了其他界面
        let noQwer = true
        let xwz = [
            AbilityButtonUtil.onlySetButtons[0], AbilityButtonUtil.onlySetButtons[1],
            AbilityButtonUtil.onlySetButtons[2], AbilityButtonUtil.onlySetButtons[3]
        ]
        let frameSelection = selection()
        //当前帧 隐藏所有按钮
        if (frameSelection != AbilityButtonUtil.lastFrameSelection) {
            AbilityButtonUtil._sl_hideBottomButton();
            AbilityButtonUtil.lastFrameSelection = frameSelection
            return
        }
        AbilityButtonUtil.lastFrameSelection = frameSelection
        let add: number[] = []
        let noAbility = true;//没有技能 选择编队时也会没有技能 选择魔法书时最下排只有关闭按钮 也没有技能
        for (let index = 0; index < 4; index++) {
            const [skid] = button(index, 2);
            if (skid == 0 || !skid) {
                continue
            }
            noAbility = false;
            const uiid = AbilityButtonUtil._sl_bottomButtonArray[index]
            /**
             * 注意单位 不能拥有相同位置的技能 否则会重叠按钮
             */
            const realButtonX = AbilityButtonUtil.abilityIdButtonXMap[skid];
            if (realButtonX) {
                xwz[realButtonX] = true
                noQwer = false;
                DzFrameSetPoint(uiid, 8, DzGetGameUI(), 8, realButtonX * 0.0435 - 0.144, 0.007)
            } else {
                add.push(uiid)
            }
        }
        //无技能时  只隐藏一帧以防止图标闪动 (防止选择魔法书类型时 隐藏了关闭按钮)
        if (noAbility && !AbilityButtonUtil.lastNoAbility) {
            AbilityButtonUtil._sl_hideBottomButton();
            AbilityButtonUtil.lastNoAbility = noAbility;
            return;
        }
        AbilityButtonUtil.lastNoAbility = noAbility;
        //设置未设置技能位置
        for (const uiid of add) {
            let bol = true
            for (let index = 0; index < xwz.length; index++) {
                if (!xwz[index]) {
                    bol = false
                    xwz[index] = true
                    DzFrameSetPoint(uiid, 8, DzGetGameUI(), 8, index * 0.0435 - 0.144, 0.007)
                    break
                }
            }
            if (bol) {
                break
            }
        }
        if (noQwer) {
            AbilityButtonUtil._sl_resetBottomButton();
        }

    }

    private static _sl_hideBottomButton() {
        for (let index = 0; index < 4; index++) {
            const uiid = AbilityButtonUtil._sl_bottomButtonArray[index]
            DzFrameSetPoint(uiid, 8, DzGetGameUI(), 8, -1, -1)
        }

    }

    //重置UI
    private static _sl_resetBottomButton() {
        for (let index = 0; index < 4; index++) {
            const uiid = AbilityButtonUtil._sl_bottomButtonArray[index]
            DzFrameSetPoint(uiid, 8, DzGetGameUI(), 8, index * 0.0435 - 0.144, 0.007)
        }
    }

    /**
     * addons
     */

    /**
     * 根据按钮的序号 1-12
     * 从左到右 从上到下 左上角为1(0,0) 右下角为12(3,2)
     * 获得 技能原始xy坐标
     */
    static getPosByNumber(number: Num_1_12 | -1 | number): Vector {
        if (number < 1 || number > 12) {
            return {x: 0, y: -11}
        }
        let x = (number - 1) % 4
        let y = Math.floor((number - 1) / 4)
        return {x: x, y: y}
    }


    /**
     * 根据按钮的xy 获得1-12的编号
     */
    static getNumberByPos(pos: Vector): number
    static getNumberByPos(x: number, y: number): number
    static getNumberByPos(xOrPos: Vector | number, y?: number): number {
        if (xOrPos == null) {
            return 0;
        }
        let x = 0;
        if (y == null) {
            x = (xOrPos as Vector).x
            y = (xOrPos as Vector).y
        } else {
            x = xOrPos as number
        }
        return y * 4 + x + 1;
    }

    /**
     * 根据宽松格式的pos 获得xy结构的pos
     */
    static getPos(num: number): Vector
    static getPos(pos: Vector): Vector
    static getPos(x: number, y: number): Vector
    static getPos(xOrPosOrNum: any, y?: number): Vector {
        let x = 0;
        if (y == null) {
            if (LangUtil.isNumber(xOrPosOrNum)) {
                if (xOrPosOrNum == -1) {
                    return {x: 0, y: -11}
                }
                let x = (xOrPosOrNum - 1) % 4
                let y = Math.floor((xOrPosOrNum - 1) / 4)
                return {x: x, y: y}
            } else {
                return xOrPosOrNum as Vector
            }
        } else {
            x = xOrPosOrNum as number
        }
        return {x, y};
    }


    /**
     * 根据屏幕位置获取技能的xy
     * (改过ui的可以重写此函数以兼容调用此函数的系统 AbilityButtonUtil.getPosBySceneXY = xxx)
     * @param sceneX
     * @param sceneY
     */
    static getPosBySceneXY(sceneX: number, sceneY: number) {
        let rY = 2 - Math.floor((sceneY - 0.01) / 0.041);
        if (rY < 0 || rY > 2) {
            return null;
        }
        let rX = Math.floor((sceneX - 0.618) / 0.0425);
        if (rX < 0 || rX > 3) {
            return null;
        }
        return {x: rX, y: rY};
    }


    static btnUIWidth = 0.04
    static btnUIHeight = 0.04
    static btnUIXGap = 0.0035
    static btnUIYGap = 0.0035
    static btnUIXMargin = 0.012
    static btnUIYMargin = 0.007

    /**
     * 根据技能图标按钮位置获得ui的xy位置 (描点为右下角对GameUI的右下角)
     * 使用示例:
     let uiPos = AbilityButtonUtil.getUIXYByPos(0, 0);
     DzFrameSetAbsolutePoint(ui, 8 , uiPos.x, uiPos.y)
     * @param x
     * @param y
     */
    static getUIXYByPos(x: number, y: number): Vector {
        let uiX = 0.8 - (3 - x) * (AbilityButtonUtil.btnUIWidth + AbilityButtonUtil.btnUIXGap) - AbilityButtonUtil.btnUIXMargin
        let uiY = (2 - y) * (AbilityButtonUtil.btnUIHeight + AbilityButtonUtil.btnUIYGap) + AbilityButtonUtil.btnUIYMargin
        return {x: uiX, y: uiY}
    }

}

File: scripts/solar/solar-common/util/ability/AbilityButtonUtil.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local __TS__SourceMapTraceBack = ____lualib.__TS__SourceMapTraceBack
__TS__SourceMapTraceBack(debug.getinfo(1).short_src, {["6"] = 1,["7"] = 1,["8"] = 2,["9"] = 2,["10"] = 3,["11"] = 3,["12"] = 4,["13"] = 4,["14"] = 8,["15"] = 8,["16"] = 8,["18"] = 8,["19"] = 49,["20"] = 50,["21"] = 51,["22"] = 52,["23"] = 53,["24"] = 54,["25"] = 49,["26"] = 66,["27"] = 67,["28"] = 68,["29"] = 69,["30"] = 70,["31"] = 71,["33"] = 73,["34"] = 74,["37"] = 77,["38"] = 78,["39"] = 66,["40"] = 83,["41"] = 84,["44"] = 87,["45"] = 87,["46"] = 87,["47"] = 87,["48"] = 87,["49"] = 87,["51"] = 93,["52"] = 93,["53"] = 94,["54"] = 95,["55"] = 93,["58"] = 97,["59"] = 98,["60"] = 99,["61"] = 83,["62"] = 105,["63"] = 107,["64"] = 108,["65"] = 112,["66"] = 114,["67"] = 115,["68"] = 116,["71"] = 119,["72"] = 120,["73"] = 121,["75"] = 122,["76"] = 122,["78"] = 123,["79"] = 124,["80"] = 125,["82"] = 127,["83"] = 128,["85"] = 132,["86"] = 133,["87"] = 134,["88"] = 135,["89"] = 136,["90"] = 136,["91"] = 136,["92"] = 136,["93"] = 136,["94"] = 136,["95"] = 136,["96"] = 136,["98"] = 138,["102"] = 122,["105"] = 142,["106"] = 143,["107"] = 144,["110"] = 147,["111"] = 149,["112"] = 150,["114"] = 151,["115"] = 151,["116"] = 152,["117"] = 153,["118"] = 154,["119"] = 155,["120"] = 155,["121"] = 155,["122"] = 155,["123"] = 155,["124"] = 155,["125"] = 155,["126"] = 155,["129"] = 151,["132"] = 159,["136"] = 163,["137"] = 164,["139"] = 105,["140"] = 169,["142"] = 170,["143"] = 170,["144"] = 171,["145"] = 172,["146"] = 172,["147"] = 172,["148"] = 172,["149"] = 172,["150"] = 172,["151"] = 172,["152"] = 172,["153"] = 170,["156"] = 169,["157"] = 178,["159"] = 179,["160"] = 179,["161"] = 180,["162"] = 181,["163"] = 181,["164"] = 181,["165"] = 181,["166"] = 181,["167"] = 181,["168"] = 181,["169"] = 181,["170"] = 179,["173"] = 178,["174"] = 194,["175"] = 195,["176"] = 196,["178"] = 198,["179"] = 199,["180"] = 200,["181"] = 194,["182"] = 209,["183"] = 210,["184"] = 211,["186"] = 213,["187"] = 214,["188"] = 215,["189"] = 216,["191"] = 218,["193"] = 220,["194"] = 209,["195"] = 229,["196"] = 230,["197"] = 231,["198"] = 232,["199"] = 233,["200"] = 234,["202"] = 236,["203"] = 237,["204"] = 238,["206"] = 240,["209"] = 243,["211"] = 245,["212"] = 229,["213"] = 255,["214"] = 256,["215"] = 257,["216"] = 258,["218"] = 260,["219"] = 261,["220"] = 262,["222"] = 264,["223"] = 255,["224"] = 283,["225"] = 284,["226"] = 285,["227"] = 286,["228"] = 283,["229"] = 10,["230"] = 11,["231"] = 19,["232"] = 20,["233"] = 31,["234"] = 38,["235"] = 81,["236"] = 102,["237"] = 103,["238"] = 268,["239"] = 269,["240"] = 270,["241"] = 271,["242"] = 272,["243"] = 273});
local ____exports = {}
local ____FrameCallbackUtil = require("solar.solar-common.util.frame.FrameCallbackUtil")
local FrameCallbackUtil = ____FrameCallbackUtil.default
local ____AbilityUtil = require("solar.solar-common.util.ability.AbilityUtil")
local AbilityUtil = ____AbilityUtil.default
local ____KeyCode = require("solar.solar-common.constant.KeyCode")
local KeyCode = ____KeyCode.default
local ____LangUtil = require("solar.solar-common.util.lang.LangUtil")
local LangUtil = ____LangUtil.default
____exports.default = __TS__Class()
local AbilityButtonUtil = ____exports.default
AbilityButtonUtil.name = "AbilityButtonUtil"
function AbilityButtonUtil.prototype.____constructor(self)
end
function AbilityButtonUtil.setAbilityBottomButtonXAndHotKey(self, unit, abilityIdStr, x)
    local player = GetOwningPlayer(unit)
    local keyCode = ____exports.default.qwerKeyCodes[x + 1]
    AbilityUtil:setUnitAbilityHotkey(unit, abilityIdStr, keyCode, false)
    AbilityUtil:setUnitAbilityName(unit, abilityIdStr, ((_g_objs.ability[abilityIdStr].Name .. "(") .. ____exports.default.qwerKeyNames[x + 1]) .. ")", true)
    return ____exports.default:setAbilityBottomButtonX(abilityIdStr, x, player)
end
function AbilityButtonUtil.setAbilityBottomButtonX(self, abilityIdStr, x, player)
    abilityIdStr = LangUtil:getIntId(abilityIdStr)
    local sABXMPD = ____exports.default.syncAbilityIdButtonXMap[GetPlayerId(player)]
    if not sABXMPD then
        sABXMPD = {}
        ____exports.default.syncAbilityIdButtonXMap[GetPlayerId(player)] = sABXMPD
    end
    sABXMPD[abilityIdStr] = x
    if player ~= GetLocalPlayer() then
        return
    end
    ____exports.default:_sl_init_bottomButton_system()
    ____exports.default.abilityIdButtonXMap[abilityIdStr] = x
end
function AbilityButtonUtil._sl_init_bottomButton_system(self)
    if ____exports.default._sl_isInitialized then
        return
    end
    ____exports.default._sl_bottomButtonArray = {
        DzFrameGetCommandBarButton(2, 0),
        DzFrameGetCommandBarButton(2, 1),
        DzFrameGetCommandBarButton(2, 2),
        DzFrameGetCommandBarButton(2, 3)
    }
    do
        local index = 0
        while index < 4 do
            local uiid = ____exports.default._sl_bottomButtonArray[index + 1]
            DzFrameClearAllPoints(uiid)
            index = index + 1
        end
    end
    ____exports.default:_sl_resetBottomButton()
    FrameCallbackUtil:addFrameSetUpdateCallback(____exports.default._sl_refreshButton)
    ____exports.default._sl_isInitialized = true
end
function AbilityButtonUtil._sl_refreshButton()
    local noQwer = true
    local xwz = {____exports.default.onlySetButtons[1], ____exports.default.onlySetButtons[2], ____exports.default.onlySetButtons[3], ____exports.default.onlySetButtons[4]}
    local frameSelection = selection()
    if frameSelection ~= ____exports.default.lastFrameSelection then
        ____exports.default:_sl_hideBottomButton()
        ____exports.default.lastFrameSelection = frameSelection
        return
    end
    ____exports.default.lastFrameSelection = frameSelection
    local add = {}
    local noAbility = true
    do
        local index = 0
        while index < 4 do
            do
                local skid = button(index, 2)
                if skid == 0 or not skid then
                    goto __continue12
                end
                noAbility = false
                local uiid = ____exports.default._sl_bottomButtonArray[index + 1]
                --- 注意单位 不能拥有相同位置的技能 否则会重叠按钮
                local realButtonX = ____exports.default.abilityIdButtonXMap[skid]
                if realButtonX then
                    xwz[realButtonX + 1] = true
                    noQwer = false
                    DzFrameSetPoint(
                        uiid,
                        8,
                        DzGetGameUI(),
                        8,
                        realButtonX * 0.0435 - 0.144,
                        0.007
                    )
                else
                    add[#add + 1] = uiid
                end
            end
            ::__continue12::
            index = index + 1
        end
    end
    if noAbility and not ____exports.default.lastNoAbility then
        ____exports.default:_sl_hideBottomButton()
        ____exports.default.lastNoAbility = noAbility
        return
    end
    ____exports.default.lastNoAbility = noAbility
    for ____, uiid in ipairs(add) do
        local bol = true
        do
            local index = 0
            while index < #xwz do
                if not xwz[index + 1] then
                    bol = false
                    xwz[index + 1] = true
                    DzFrameSetPoint(
                        uiid,
                        8,
                        DzGetGameUI(),
                        8,
                        index * 0.0435 - 0.144,
                        0.007
                    )
                    break
                end
                index = index + 1
            end
        end
        if bol then
            break
        end
    end
    if noQwer then
        ____exports.default:_sl_resetBottomButton()
    end
end
function AbilityButtonUtil._sl_hideBottomButton(self)
    do
        local index = 0
        while index < 4 do
            local uiid = ____exports.default._sl_bottomButtonArray[index + 1]
            DzFrameSetPoint(
                uiid,
                8,
                DzGetGameUI(),
                8,
                -1,
                -1
            )
            index = index + 1
        end
    end
end
function AbilityButtonUtil._sl_resetBottomButton(self)
    do
        local index = 0
        while index < 4 do
            local uiid = ____exports.default._sl_bottomButtonArray[index + 1]
            DzFrameSetPoint(
                uiid,
                8,
                DzGetGameUI(),
                8,
                index * 0.0435 - 0.144,
                0.007
            )
            index = index + 1
        end
    end
end
function AbilityButtonUtil.getPosByNumber(self, number)
    if number < 1 or number > 12 then
        return {x = 0, y = -11}
    end
    local x = (number - 1) % 4
    local y = math.floor((number - 1) / 4)
    return {x = x, y = y}
end
function AbilityButtonUtil.getNumberByPos(self, xOrPos, y)
    if xOrPos == nil then
        return 0
    end
    local x = 0
    if y == nil then
        x = xOrPos.x
        y = xOrPos.y
    else
        x = xOrPos
    end
    return y * 4 + x + 1
end
function AbilityButtonUtil.getPos(self, xOrPosOrNum, y)
    local x = 0
    if y == nil then
        if LangUtil:isNumber(xOrPosOrNum) then
            if xOrPosOrNum == -1 then
                return {x = 0, y = -11}
            end
            local x = (xOrPosOrNum - 1) % 4
            local y = math.floor((xOrPosOrNum - 1) / 4)
            return {x = x, y = y}
        else
            return xOrPosOrNum
        end
    else
        x = xOrPosOrNum
    end
    return {x = x, y = y}
end
function AbilityButtonUtil.getPosBySceneXY(self, sceneX, sceneY)
    local rY = 2 - math.floor((sceneY - 0.01) / 0.041)
    if rY < 0 or rY > 2 then
        return nil
    end
    local rX = math.floor((sceneX - 0.618) / 0.0425)
    if rX < 0 or rX > 3 then
        return nil
    end
    return {x = rX, y = rY}
end
function AbilityButtonUtil.getUIXYByPos(self, x, y)
    local uiX = 0.8 - (3 - x) * (____exports.default.btnUIWidth + ____exports.default.btnUIXGap) - ____exports.default.btnUIXMargin
    local uiY = (2 - y) * (____exports.default.btnUIHeight + ____exports.default.btnUIYGap) + ____exports.default.btnUIYMargin
    return {x = uiX, y = uiY}
end
AbilityButtonUtil.qwerKeyCodes = {KeyCode.VK_Q, KeyCode.VK_W, KeyCode.VK_E, KeyCode.VK_R}
AbilityButtonUtil.qwerKeyNames = {"Q", "W", "E", "R"}
AbilityButtonUtil.onlySetButtons = {false, false, false, false}
AbilityButtonUtil._sl_bottomButtonArray = nil
AbilityButtonUtil.abilityIdButtonXMap = {}
AbilityButtonUtil.syncAbilityIdButtonXMap = {}
AbilityButtonUtil._sl_isInitialized = false
AbilityButtonUtil.lastFrameSelection = nil
AbilityButtonUtil.lastNoAbility = false
AbilityButtonUtil.btnUIWidth = 0.04
AbilityButtonUtil.btnUIHeight = 0.04
AbilityButtonUtil.btnUIXGap = 0.0035
AbilityButtonUtil.btnUIYGap = 0.0035
AbilityButtonUtil.btnUIXMargin = 0.012
AbilityButtonUtil.btnUIYMargin = 0.007
return ____exports
